lib/money/money.rb

```diff
@@ -463,49 +464,27 @@ def as_euro
     exchange_to("EUR")
   end
 
-  # Allocates money between different parties without losing pennies.
-  # After the mathematical split has been performed, leftover pennies will
-  # be distributed round-robin amongst the parties. This means that parties
-  # listed first will likely receive more pennies than ones that are listed later
+  # Splits a given amount in parts without loosing pennies. The left-over pennies will be
+  # distributed round-robin amongst the parties. This means that parties listed first will likely
+  # receive more pennies than ones that are listed later.
   #
-  # @param [Array<Numeric>] splits [2, 1, 1] to give twice as much to party1 as party2 or party3
-  #   which results in 50% of the cash to party1, 25% to party2, and 25% to party3.
+  # @param [Array<Numeric>, Numeric] pass [2, 1, 1] to give twice as much to party1 as party2 or
+  # party3 which results in 50% of the cash to party1, 25% to party2, and 25% to party3. Passing a
+  # number instead of an array will split the amount evenly (without loosing pennies when rounding).
   #
   # @return [Array<Money>]
   #
   # @example
-  #   Money.new(5,   "USD").allocate([0.3, 0.7])         #=> [Money.new(2), Money.new(3)]
+  #   Money.new(5,   "USD").allocate([0.3, 0.7]) #=> [Money.new(2), Money.new(3)]
   #   Money.new(100, "USD").allocate([1, 1, 1]) #=> [Money.new(34), Money.new(33), Money.new(33)]
+  #   Money.new(100, "USD").allocate(2) #=> [Money.new(50), Money.new(50)]
+  #   Money.new(100, "USD").allocate(3) #=> [Money.new(34), Money.new(33), Money.new(33)]
   #
-  def allocate(splits)
-    amounts, left_over = amounts_from_splits(splits)
-
-    unless self.class.infinite_precision
-      delta = left_over > 0 ? 1 : -1
-      # Distribute left over pennies amongst allocations
-      left_over.to_i.abs.times { |i| amounts[i % amounts.length] += delta }
-    end
-
-    amounts.collect { |fractional| self.class.new(fractional, currency) }
-  end
-
-  # Split money amongst parties evenly without losing pennies.
-  #
-  # @param [Numeric] num number of parties.
-  #
-  # @return [Array<Money>]
-  #
-  # @example
-  #   Money.new(100, "USD").split(3) #=> [Money.new(34), Money.new(33), Money.new(33)]
-  def split(num)
-    raise ArgumentError, "need at least one party" if num < 1
-
-    if self.class.infinite_precision
-      split_infinite(num)
-    else
-      split_flat(num)
-    end
+  def allocate(parts)
```

Comment:
WDYT about extracting this methods to `Money::Allocate` and including it into Money? With @ct-clearhaus 's suggestions `generate` becomes clean and straightforward, so it can be used without instantiating an object and be implemented as private method or class method of new module?

Reply:
I see this as inheritance vs composition decision. While there's a certain appeal in including this directly into Money, I don't think there's much benefit in doing so — object instantiation is cheap in ruby (plus it's just 1 object) and it's easier to add more complex behaviour (like switchable distribution) to a separate class.

Happy to discuss this further if you feel like I'm missing something here.

Reply:
Here is what I suggest:

```ruby
class Money
  module Allocation
    class << self
     def split(amount, parts, ...)
    end

    def allocate(...)
      Allocate.split(...)
    end

    alias_method :split, :allocate
  end
end

# in lib/money.rb
class Money
  include Allocation
end
```

This way allocation methods are in separate module, because it's like an extended. Similar to `Arithmetics`. If you want you can keep a class `Money::Allocation::Generator`, but for now I see no pros in using it vs a static method.

Reply:
Don't you think this is creating a tight coupling between the two? As of the code in the PR — `Money::Allocation` does not depend on Money in any way. Some obvious benefits I see are: 

- being able to easier test it in isolation (no need to instantiate all the Money objects to check results)
- `Allocation` is not concerned with aliasing the method for backwards-compatibility (which I think it shouldn't)
- The public method is defined on the `Money` directly, making it easier to find (less indirection)

Correct me if I'm wrong, but what you are suggesting is essentially splitting the code between different files without adding a class boundary. Why do you think it's a better approach?

Reply:
1) It's even easier, because there is no need to instantiate `Allocation` as it's now. Just `Money::Allocation.split`, similar to testing my example from gist.
2, 3) This is just collecting methods together by functionality. Why there is separate Arithmetics module? Then let's move all methods to Money.

I just want to keep different functionality separate from other. Money should include only the basics. Allocations are not essential for most of users, so related methods must be kept separately. I use this test: what should i do to remove functionality? With module I need just remove include and delete file, without - look for all related methods and files and remove them one by one.

Reply:
+1 for "minimality".

Reply:
@printercu I see the reason for moving to class methods, I'll make this change.

As for grouping methods by functionality — yes, sometimes there is a need to do so and Arithmetics is a good example, but it also indicates that the public interface for this class is too big. In case you're familiar with SOLID design principles, it's the violation of the Interface Segregation principle — you're making collaborators depend on a bunch of stuff they don't need/use.

Normally OO design is judged by how easy it is to change things. In your suggestion the `Money` depends on `Allocation`, as well as `Allocation` depends on `Money` — adding more reasons for both to change.

I suggest we agree to disagree here. However I would appreciate some input on things that matter — see the comment at the bottom of this PR.